### ISSUE TITLE
Soes LanceDB support concurrent writes to S3?

### ISSUE THREAD

**Issue by mrocklin** [lancedb/lancedb#1077]
### SDK

Python

### Description

I'm guessing not, but wanted to see what the recommendation was for large scale writes in a serverless cloud environment.

**Comment by westonpace** [comment]
S3 doesn't support atomic renames (well, to be specific, something like "rename if not exists") which is what we use for concurrency.  So, out of the box, it does not.

Lance (the table format that lancedb is built on) does have some advanced options for using alternative commit mechanisms.  For example, we have some S3 deployments using dynamodb commit locking.  You can read more about it here: https://lancedb.github.io/lance/read_and_write.html#committing-mechanisms-for-s3

This is something we would like to bring to lancedb at some point, and it may work for lancedb today with advanced configuration, but it's not something I have done much testing with.

**Comment by westonpace** [comment]
This is probably related: https://github.com/lancedb/lancedb/issues/534

**Comment by mrocklin** [comment]
Let's say I have a Python `threading.Lock` equivalent object that happens to work in a distributed manner. Right now I think that my best option here is to do something like this:

```python
def process(filename):
    db = lancedb.connect("s3://my-bucket/my-lance-db")
    tbl = db.open_table("my_table")

    df = load(filename)
    df = calculate_embeddings(df)

    with lock:
        tbl.add(df)

distributed_map(process, filenames)
```

Would that work?  To be clear, what I'm asking is if many distributed processes can write to the same s3-backed database if I guarantee that they don't write concurrently, but that one might write milliseconds after another finishes.  This would not work if there was state within the process itself that was needed in order to ensure consistency.

**Comment by westonpace** [comment]
Yes, that should work fine.  There will be other operations (update / delete / etc.) that will need to be guarded with the lock but it should be fairly clear which they are.  Concurrent reads and reads concurrent with writes should be fine as well.

**Comment by mrocklin** [comment]
Cool. Thank you.  This answers my question.  I'm going to go ahead and close this (but I'm also happy if you want to reopen for future feature request). My apologies for using github to ask a question rather than discord (I tend to avoid chat systems).
